File: 键盘监听/键盘监听/ViewController.m
//
//  ViewController.m
//  键盘监听
//
//  Created by Lone on 16/6/2.
//  Copyright © 2016年 MAC. All rights reserved.
//

#import "ViewController.h"

@interface ViewController ()

@end

@implementation ViewController

- (void)touchesBegan:(NSSet<UITouch *> *)touches withEvent:(UIEvent *)event
{
    [self.view endEditing:YES];
}

- (void)viewWillAppear:(BOOL)animated
{
    [[NSNotificationCenter defaultCenter] addObserver:self
                                             selector:@selector(keyboardWillShow:)
                                                 name:UIKeyboardWillShowNotification
                                               object:nil];
    [[NSNotificationCenter defaultCenter] addObserver:self
                                             selector:@selector(keyboardWillHide:)
                                                 name:UIKeyboardWillHideNotification
                                               object:nil];
}

- (void)viewDidLoad {
    [super viewDidLoad];
//    [[NSNotificationCenter defaultCenter] addObserver:self
//                                             selector:@selector(keyboardWillShow:)
//                                                 name:UIKeyboardWillShowNotification
//                                               object:nil];
//    [[NSNotificationCenter defaultCenter] addObserver:self
//                                             selector:@selector(keyboardWillHide:)
//                                                 name:UIKeyboardWillHideNotification
//                                               object:nil];
    // Do any additional setup after loading the view, typically from a nib.
}

- (void)keyboardWillShow:(NSNotification *)nf
{
    //获取键盘的高度
    
    NSDictionary *userInfo = [nf userInfo];
    
    NSValue *endValue = [userInfo objectForKey:UIKeyboardFrameEndUserInfoKey];
    NSValue *beginValue = [userInfo objectForKey:UIKeyboardFrameBeginUserInfoKey];
    
    CGRect keyboardRect_end = [endValue CGRectValue];
    CGRect keyboardRect_begin = [beginValue CGRectValue];
    
    int height = keyboardRect_end.size.height;
    
    NSLog(@"userInfo---------------%@",userInfo);
    
    NSLog(@"---------------%d",height);
    
    NSLog(@"keyboardRect_begin.size.height---------------%d",(int)keyboardRect_begin.size.height);
    
    if (keyboardRect_begin.size.height > 0) {
        [UIView animateWithDuration:0.25 animations:^{
            CGRect frame = self.view.frame;
            frame.origin.y = - height;
            self.view.frame = frame;
        }];
    }
    
}

- (void)keyboardWillHide:(NSNotification *)nf
{
    [UIView animateWithDuration:0.25 animations:^{
        CGRect frame = self.view.frame;
        frame.origin.y = 0;
        self.view.frame = frame;
    }];
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

@end

File: 李健凡iOS知识总结 /swift仿微信键盘变化/swift仿微信键盘变化/ViewController.swift
//
//  ViewController.swift
//  swift仿微信键盘变化
//
//  Created by mini on 15/10/20.
//  Copyright © 2015年 mini. All rights reserved.
//

import UIKit

class ViewController: UIViewController {

    @IBOutlet weak var keyBoardView: UITextField!
    override func viewDidLoad() {
        super.viewDidLoad()
        
        NSNotificationCenter.defaultCenter().addObserver(self, selector: "keyBoardWillShow:", name: UIKeyboardWillShowNotification, object: nil)
        @IBOutlet weak var board: Board!
        NSNotificationCenter.defaultCenter().addObserver(self, selector: "keyBoardWillhide:", name: UIKeyboardWillHideNotification, object: nil)
        // Do any additional setup after loading the view, typically from a nib.
    }
    
    override func touchesBegan(touches: Set<UITouch>, withEvent event: UIEvent?) {
        self.keyBoardView.resignFirstResponder()
    }
    
    func keyBoardWillShow(note:NSNotification)
    {
        //1.将通知的用户信息取出,转化为字典类型
        let userInfo = note.userInfo as! NSDictionary
        //2.通过对应的键UIKeyboardFrameEndUserInfoKey，取出键盘位置信息
        let keyBoardBounds = (userInfo[UIKeyboardFrameEndUserInfoKey] as! NSValue).CGRectValue()
        //3.通过UIKeyboardAnimationDurationUserInfoKey,取出动画时长信息
        let duration = (userInfo[UIKeyboardAnimationDurationUserInfoKey] as! NSNumber).doubleValue
        //4.由于取出的位置信息是绝对的，所以要将其转换为对应于当前view的位置，否则位置信息会出错！
//        var keyBoardBoudsRect = self.view.convertRect(keyBoardBounds, toView: nil)
//        
//        var keyBoardViewFrame = keyBoardView.frame
        let deltaY = keyBoardBounds.size.height
        
        let animations :(() -> Void) = {
            
            //使输入框的高度减少deltaY也就是跟随键盘的位置向上移动;
            self.keyBoardView.transform = CGAffineTransformMakeTranslation(0, -deltaY)
        }
        
        if duration > 0 {
            
            //这里是将时间曲线信息(一个64为的无符号整型)转换为UIViewAnimationOptions类型，要通过左移16来完成类型转换。
            let options = UIViewAnimationOptions(rawValue: UInt((userInfo[UIKeyboardAnimationCurveUserInfoKey] as!NSNumber).integerValue << 16))
            UIView.animateWithDuration(duration, delay: 0, options: options, animations: animations, completion: nil)
        }else {
            animations()
        }
        
    }
    
    func keyBoardWillhide(note:NSNotification)
    {
        let userInfo = note.userInfo as! NSDictionary
        let durarion = (userInfo[UIKeyboardAnimationDurationUserInfoKey] as! NSNumber).doubleValue
        let animations :(() -> Void) = {
            self.keyBoardView.transform = CGAffineTransformIdentity
        }
        if durarion > 0 {
            let options = UIViewAnimationOptions(rawValue: UInt((userInfo[UIKeyboardAnimationCurveUserInfoKey] as! NSNumber).integerValue << 16))
            UIView.animateWithDuration(durarion, delay: 0, options:options, animations: animations, completion: nil)
        }else{
            animations()
        }
    }
    
    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }


}

